File: 5_qubit_code/5_qubit_decoder.py
import sys

# Global flag for phase measurement
flag = False

# Syndrome-to-correction dictionaries for XZZXI
DICT1_TRUE = [
    {"syndrome": [1, 1, 1, 1],   "correction": "IIIII"},
    {"syndrome": [-1, -1, 1, 1],  "correction": "IXZXI"},
    {"syndrome": [-1, 1, 1, -1],  "correction": "IYZXI"},
    {"syndrome": [1, 1, 1, -1],   "correction": "IZZXI"},
    {"syndrome": [1, -1, -1, 1],  "correction": "IIIXI"},
    {"syndrome": [-1, 1, -1, 1],  "correction": "IIXXI"},
    {"syndrome": [-1, 1, 1, 1],   "correction": "IIYXI"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIZXI"},
]
DICT1_FALSE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [1, 1, 1, -1],   "correction": "XIIII"},
    {"syndrome": [-1, 1, -1, -1], "correction": "YIIII"},
    {"syndrome": [-1, 1, -1, 1],  "correction": "ZIIII"},
    {"syndrome": [-1, 1, 1, 1],   "correction": "IXIII"},
    {"syndrome": [-1, -1, 1, -1], "correction": "IYIII"},
    {"syndrome": [1, -1, 1, -1],  "correction": "IZIII"},
    {"syndrome": [-1, -1, 1, 1],  "correction": "IIXII"},
    {"syndrome": [-1, -1, -1, 1], "correction": "IIYII"},
    {"syndrome": [1, 1, -1, 1],   "correction": "IIZII"},
    {"syndrome": [1, -1, -1, 1],  "correction": "IIIXI"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IIIYI"},
    {"syndrome": [-1, 1, 1, -1],  "correction": "IIIZI"},
]

# Syndrome-to-correction dictionaries for IXZZX
DICT2_TRUE = [
    {"syndrome": [1, 1, 1, 1],   "correction": "IIIII"},
    {"syndrome": [-1, 1, -1, 1], "correction": "IIIZX"},
    {"syndrome": [1, -1, -1, 1], "correction": "IIXZX"},
    {"syndrome": [1, -1, 1, 1],  "correction": "IIYZX"},
    {"syndrome": [-1, 1, 1, 1],  "correction": "IIZZX"},
    {"syndrome": [1, 1, -1, -1], "correction": "IIIIX"},
    {"syndrome": [1, -1, 1, -1], "correction": "IIIXX"},
    {"syndrome": [-1, -1, 1, 1], "correction": "IIIYX"},
]
DICT2_FALSE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [-1, 1, 1, 1],   "correction": "IXIII"},
    {"syndrome": [-1, -1, 1, -1], "correction": "IYIII"},
    {"syndrome": [1, -1, 1, -1],  "correction": "IZIII"},
    {"syndrome": [-1, -1, 1, 1],  "correction": "IIXII"},
    {"syndrome": [-1, -1, -1, 1], "correction": "IIYII"},
    {"syndrome": [1, 1, -1, 1],   "correction": "IIZII"},
    {"syndrome": [1, -1, -1, 1],  "correction": "IIIXI"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IIIYI"},
    {"syndrome": [-1, 1, 1, -1],  "correction": "IIIZI"},
    {"syndrome": [1, 1, -1, -1],  "correction": "IIIIX"},
    {"syndrome": [1, -1, -1, -1], "correction": "IIIIY"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIIIZ"},
]

# Syndrome-to-correction dictionaries for XIXZZ
DICT3_TRUE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [-1, -1, 1, -1], "correction": "IIIZZ"},
    {"syndrome": [1, 1, 1, -1],   "correction": "IIXZZ"},
    {"syndrome": [1, 1, -1, -1],  "correction": "IIYZZ"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IIZZZ"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIIIZ"},
    {"syndrome": [1, 1, -1, 1],   "correction": "IIIXZ"},
    {"syndrome": [-1, 1, -1, -1], "correction": "IIIYZ"},
]
DICT3_FALSE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [1, 1, 1, -1],   "correction": "XIIII"},
    {"syndrome": [-1, 1, -1, -1], "correction": "YIIII"},
    {"syndrome": [-1, 1, -1, 1],  "correction": "ZIIII"},
    {"syndrome": [-1, -1, 1, 1],  "correction": "IIXII"},
    {"syndrome": [-1, -1, -1, 1], "correction": "IIYII"},
    {"syndrome": [1, 1, -1, 1],   "correction": "IIZII"},
    {"syndrome": [1, -1, -1, 1],  "correction": "IIIXI"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IIIYI"},
    {"syndrome": [-1, 1, 1, -1],  "correction": "IIIZI"},
    {"syndrome": [1, 1, -1, -1],  "correction": "IIIIX"},
    {"syndrome": [1, -1, -1, -1], "correction": "IIIIY"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIIIZ"},
]

# Syndrome-to-correction dictionaries for ZXIXZ
DICT4_TRUE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [1, 1, -1, 1],   "correction": "IIIXZ"},
    {"syndrome": [-1, 1, -1, 1],  "correction": "IXIXZ"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IYIXZ"},
    {"syndrome": [1, -1, -1, -1], "correction": "IZIXZ"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIIIZ"},
    {"syndrome": [-1, 1, -1, -1], "correction": "IIIYZ"},
    {"syndrome": [-1, -1, 1, -1], "correction": "IIIZZ"},
]
DICT4_FALSE = [
    {"syndrome": [1, 1, 1, 1],    "correction": "IIIII"},
    {"syndrome": [1, 1, 1, -1],   "correction": "XIIII"},
    {"syndrome": [-1, 1, -1, -1], "correction": "YIIII"},
    {"syndrome": [-1, 1, -1, 1],  "correction": "ZIIII"},
    {"syndrome": [-1, 1, 1, 1],   "correction": "IXIII"},
    {"syndrome": [-1, -1, 1, -1], "correction": "IYIII"},
    {"syndrome": [1, -1, 1, -1],  "correction": "IZIII"},
    {"syndrome": [1, -1, -1, 1],  "correction": "IIIXI"},
    {"syndrome": [-1, -1, -1, -1],"correction": "IIIYI"},
    {"syndrome": [-1, 1, 1, -1],  "correction": "IIIZI"},
    {"syndrome": [1, 1, -1, -1],  "correction": "IIIIX"},
    {"syndrome": [1, -1, -1, -1], "correction": "IIIIY"},
    {"syndrome": [1, -1, 1, 1],   "correction": "IIIIZ"},
]

# Lookup utility
def lookup_correction(syndrome, true_dict, false_dict):
    table = true_dict if flag else false_dict
    for entry in table:
        if entry["syndrome"] == syndrome:
            return entry["correction"]
    return None

# Syndrome measurement functions
# XZZXI
def syndrome1(syndrome):
    return lookup_correction(syndrome, DICT1_TRUE, DICT1_FALSE)

# IXZZX
def syndrome2(syndrome):
    return lookup_correction(syndrome, DICT2_TRUE, DICT2_FALSE)

# XIXZZ
def syndrome3(syndrome):
    return lookup_correction(syndrome, DICT3_TRUE, DICT3_FALSE)

# ZXIXZ
def syndrome4(syndrome):
    return lookup_correction(syndrome, DICT4_TRUE, DICT4_FALSE)

File: 5_qubit_code/5_qubit_decoder.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdbool.h>

bool flag;
struct dictionary
{
    int syndrome[4];
    char *correction;
};

char *syndrome1(bool flag);
char *syndrome2(bool flag);
char *syndrome3(bool flag);
char *syndrome4(bool flag);

int main()
{
    int n;
    scanf("%d", &n);
    if (n == 1)
        flag = false;
    else if (n == -1)
        flag = true;
    else
    {
        printf("Incorrect Flag Measurement.\n");
        exit(1);
    }

    char *correction1, *correction2, *correction3, *correction4; 
    if ((correction1 = malloc(sizeof(char) * 6)) == NULL)
    {
        printf("Correction 1 Allocation Failed.\n");
        exit(1);
    }
    if ((correction2 = malloc(sizeof(char) * 6)) == NULL)
    {
        printf("Correction 2 Allocation Failed.\n");
        exit(1);
    }
    if ((correction3 = malloc(sizeof(char) * 6)) == NULL)
    {
        printf("Correction 3 Allocation Failed.\n");
        exit(1);
    }
    if ((correction4 = malloc(sizeof(char) * 6)) == NULL)
    {
        printf("Correction 4 Allocation Failed.\n");
        exit(1);
    }

    if ((correction1 = syndrome1(flag)) == NULL)
    {
        printf("No valid correction for syndrome measurement 1.\n");
        exit(1);
    }
    if ((correction2 = syndrome2(flag)) == NULL)
    {
        printf("No valid correction for syndrome measurement 2.\n");
        exit(1);
    }
    if ((correction3 = syndrome3(flag)) == NULL)
    {
        printf("No valid correction for syndrome measurement 3.\n");
        exit(1);
    }
    if ((correction4 = syndrome4(flag)) == NULL)
    {
        printf("No valid correction for syndrome measurement 4.\n");
        exit(1);
    }

    free(correction1);
    free(correction2);
    free(correction3);
    free(correction4);
    return 0;
}

char *syndrome1(bool flag)
{
    // Measuring XZZXI
    int syndrome[4];
    int *p = syndrome;
    for (; p < syndrome + 4; p++)
    {
        scanf(" %d", p++);
        if (abs(*p) != 1)
        {
            printf("Incorrect Stabilizer Measurement Outcome.\n");
            exit(1);
        }
    }

    if (flag)
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {-1, -1, 1, 1}, .correction = "IXZXI"},
                                    [2] = { .syndrome = {-1, 1, 1, -1}, .correction = "IYZXI"},
                                    [3] = { .syndrome = {1, 1, 1, -1}, .correction = "IZZXI"},
                                    [4] = { .syndrome = {1, -1, -1, 1}, .correction = "IIIXI"},
                                    [5] = { .syndrome = {-1, 1, -1, 1}, .correction = "IIXXI"},
                                    [6] = { .syndrome = {-1, 1, 1, 1}, .correction = "IIYXI"},
                                    [7] = { .syndrome = {1, -1, 1, 1}, .correction = "IIZXI"}};

        int i, j;
        for (i = 0; i < 8; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
    else
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {1, 1, 1, -1}, .correction = "XIIII"},
                                    [2] = { .syndrome = {-1, 1, -1, -1}, .correction = "YIIII"},
                                    [3] = { .syndrome = {-1, 1, -1 ,1}, .correction = "ZIIII"},
                                    [4] = { .syndrome = {-1, 1, 1, 1}, .correction = "IXIII"},
                                    [5] = { .syndrome = {-1, -1, 1, -1}, .correction = "IYIII"},
                                    [6] = { .syndrome = {1, -1, 1, -1}, .correction = "IZIII"},
                                    [7] = { .syndrome = {-1, -1, 1, 1}, .correction = "IIXII"},
                                    [8] = { .syndrome = {-1, -1, -1, 1}, .correction = "IIYII"},
                                    [9] = { .syndrome = {1, 1, -1, 1}, .correction = "IIZII"},
                                    [10] = { .syndrome = {1, -1, -1 ,1}, .correction = "IIIXI"},
                                    [11] = { .syndrome = {-1, -1, -1, -1}, .correction = "IIIYI"},
                                    [12] = { .syndrome = {-1, 1, 1, -1}, .correction = "IIIZI"}};

        int i, j;
        for (i = 0; i < 13; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
}


char *syndrome2(bool flag)
{
    // Measuring IXZZX
    int syndrome[4];
    int *p = syndrome;
    for (; p < syndrome + 4; p++)
    {
        scanf(" %d", p++);
        if (abs(*p) != 1)
        {
            printf("Incorrect Stabilizer Measurement Outcome.\n");
            exit(1);
        }
    }

    if (flag)
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {-1, 1, -1, 1}, .correction = "IIIZX"},
                                    [2] = { .syndrome = {1, -1, -1, 1}, .correction = "IIXZX"},
                                    [3] = { .syndrome = {1, -1, 1, 1}, .correction = "IIYZX"},
                                    [4] = { .syndrome = {-1, 1, 1, 1}, .correction = "IIZZX"},
                                    [5] = { .syndrome = {1, 1, -1, -1}, .correction = "IIIIX"},
                                    [6] = { .syndrome = {1, -1, 1, -1}, .correction = "IIIXX"},
                                    [7] = { .syndrome = {-1, -1, 1, 1}, .correction = "IIIYX"}};

        int i, j;
        for (i = 0; i < 8; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
    else
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {-1, 1, 1, 1}, .correction = "IXIII"},
                                    [2] = { .syndrome = {-1, -1, 1, -1}, .correction = "IYIII"},
                                    [3] = { .syndrome = {1, -1, 1, -1}, .correction = "IZIII"},
                                    [4] = { .syndrome = {-1, -1, 1, 1}, .correction = "IIXII"},
                                    [5] = { .syndrome = {-1, -1, -1, 1}, .correction = "IIYII"},
                                    [6] = { .syndrome = {1, 1, -1, 1}, .correction = "IIZII"},
                                    [7] = { .syndrome = {1, -1, -1, 1}, .correction = "IIIXI"},
                                    [8] = { .syndrome = {-1, -1, -1, -1}, .correction = "IIIYI"},
                                    [9] = { .syndrome = {-1, 1, 1, -1}, .correction = "IIIZI"},
                                    [10] = { .syndrome = {1, 1, -1, -1}, .correction = "IIIIX"},
                                    [11] = { .syndrome = {1, -1, -1, -1}, .correction = "IIIIY"},
                                    [12] = { .syndrome = {1, -1, 1, 1}, .correction = "IIIIZ"}};

        int i, j;
        for (i = 0; i < 13; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
}


char *syndrome3(bool flag)
{
    // Measuring XIXZZ
    int syndrome[4];
    int *p = syndrome;
    for (; p < syndrome + 4; p++)
    {
        scanf(" %d", p++);
        if (abs(*p) != 1)
        {
            printf("Incorrect Stabilizer Measurement Outcome.\n");
            exit(1);
        }
    }

    if (flag)
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {-1, -1, 1, -1}, .correction = "IIIZZ"},
                                    [2] = { .syndrome = {1, 1, 1, -1}, .correction = "IIXZZ"},
                                    [3] = { .syndrome = {1, 1, -1, -1}, .correction = "IIYZZ"},
                                    [4] = { .syndrome = {-1, -1, -1, -1}, .correction = "IIZZZ"},
                                    [5] = { .syndrome = {1, -1, 1, 1}, .correction = "IIIIZ"},
                                    [6] = { .syndrome = {1, 1, -1, 1}, .correction = "IIIXZ"},
                                    [7] = { .syndrome = {-1, 1, -1, -1}, .correction = "IIIYZ"}};

        int i, j;
        for (i = 0; i < 8; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
    else
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {1, 1, 1, -1}, .correction = "XIIII"},
                                    [2] = { .syndrome = {-1, 1, -1, -1}, .correction = "YIIII"},
                                    [3] = { .syndrome = {-1, 1, -1 ,1}, .correction = "ZIIII"},
                                    [4] = { .syndrome = {-1, -1, 1, 1}, .correction = "IIXII"},
                                    [5] = { .syndrome = {-1, -1, -1, 1}, .correction = "IIYII"},
                                    [6] = { .syndrome = {1, 1, -1, 1}, .correction = "IIZII"},
                                    [7] = { .syndrome = {1, -1, -1, 1}, .correction = "IIIXI"},
                                    [8] = { .syndrome = {-1, -1, -1, -1}, .correction = "IIIYI"},
                                    [9] = { .syndrome = {-1, 1, 1, -1}, .correction = "IIIZI"},
                                    [10] = { .syndrome = {1, 1, -1, -1}, .correction = "IIIIX"},
                                    [11] = { .syndrome = {1, -1, -1, -1}, .correction = "IIIIY"},
                                    [12] = { .syndrome = {1, -1, 1, 1}, .correction = "IIIIZ"}};

        int i, j;
        for (i = 0; i < 13; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
}


char *syndrome4(bool flag)
{
    // Measuring ZXIXZ
    int syndrome[4];
    int *p = syndrome;
    for (; p < syndrome + 4; p++)
    {
        scanf(" %d", p++);
        if (abs(*p) != 1)
        {
            printf("Incorrect Stabilizer Measurement Outcome.\n");
            exit(1);
        }
    }

    if (flag)
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {1, 1, -1, 1}, .correction = "IIIXZ"},
                                    [2] = { .syndrome = {-1, 1, -1, 1}, .correction = "IXIXZ"},
                                    [3] = { .syndrome = {-1, -1, -1, -1}, .correction = "IYIXZ"},
                                    [4] = { .syndrome = {1, -1, -1, -1}, .correction = "IZIXZ"},
                                    [5] = { .syndrome = {1, -1, 1, 1}, .correction = "IIIIZ"},
                                    [6] = { .syndrome = {-1, 1, -1, -1}, .correction = "IIIYZ"},
                                    [7] = { .syndrome = {-1, -1, 1, -1}, .correction = "IIIZZ"}};

        int i, j;
        for (i = 0; i < 8; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
    else
    {
        struct dictionary dict[] = {[0] = { .syndrome = {1, 1, 1, 1}, .correction = "IIIII"},
                                    [1] = { .syndrome = {1, 1, 1, -1}, .correction = "XIIII"},
                                    [2] = { .syndrome = {-1, 1, -1, -1}, .correction = "YIIII"},
                                    [3] = { .syndrome = {-1, 1, -1, 1}, .correction = "ZIIII"},
                                    [4] = { .syndrome = {-1, 1, 1, 1}, .correction = "IXIII"},
                                    [5] = { .syndrome = {-1, -1, 1, -1}, .correction = "IYIII"},
                                    [6] = { .syndrome = {1, -1, 1, -1}, .correction = "IZIII"},
                                    [7] = { .syndrome = {1, -1, -1, 1}, .correction = "IIIXI"},
                                    [8] = { .syndrome = {-1, -1, -1, -1}, .correction = "IIIYI"},
                                    [9] = { .syndrome = {-1, 1, 1, -1}, .correction = "IIIZI"},
                                    [10] = { .syndrome = {1, 1, -1, -1}, .correction = "IIIIX"},
                                    [11] = { .syndrome = {1, -1, -1, -1}, .correction = "IIIIY"},
                                    [12] = { .syndrome = {1, -1, 1, 1}, .correction = "IIIIZ"}};

        int i, j;
        for (i = 0; i < 13; i++)
        {
            for (j = 0; j < 4; j++)
            {
                if (dict[i].syndrome[j] != syndrome[j])
                    break;
            }
            if (j == 4)
                return dict[i].correction;
        }
        return NULL;
    }
}